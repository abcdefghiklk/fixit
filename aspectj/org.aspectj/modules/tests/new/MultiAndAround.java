import org.aspectj.testing.Tester;

public class MultiAndAround {
    public static void main(String[] args) {
        exercise(new Base());
        exercise(new Derived());
        exercise(new SubBase());

        Tester.checkEventsFromFile("MultiAndAround.out");
    }

    static void exercise(IBase b) {
        Tester.event("**************************************");
        b.base1("1");
        b.base2("2");
        b.base3("3");
    }
}


interface IBase {
    void base1(String s);
    void base2(String s);
    void base3(String s);
}

interface IDerived extends IBase {
    void base1(String s);
}

class Base implements IBase {
    public void base2(String s) { Tester.event("Base.base2"); }
}

class Derived implements IDerived {
    public void base3(String s) { Tester.event("Derived.base3"); }
}

class SubBase extends Base {
    public void base3(String s) { Tester.event("SubBase.base3"); }
}

aspect Intro {
    public void IBase.base1(String s) { Tester.event("IBase.base1 from Intro"); }
    public void IBase.base3(String s) { Tester.event("IBase.base3 from Intro"); }

    public void IDerived.base1(String s) { Tester.event("IDerived.base1 from Intro"); }
    public void IDerived.base2(String s) { Tester.event("IDerived.base2 from Intro"); }
    public void IDerived.base3(String s) { Tester.event("IDerived.base3 from Intro"); }
}


aspect A {
    Object around(): call(* IBase.*(..)) {
        Tester.event("around call(IBase.*)");
        return proceed();
    }
    Object around(): call(* IBase+.*(..)) {
        Tester.event("around call(IBase+.*)");
        return proceed();
    }
    Object around(): call(* Base.*(..)) {
        Tester.event("around call(Base.*)");
        return proceed();
    }
    Object around(): execution(* Base+.*(..)) {
        Tester.event("around execution(Base+.*)");
        return proceed();
    }
}
