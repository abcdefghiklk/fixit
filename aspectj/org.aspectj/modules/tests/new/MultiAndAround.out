**************************************
around call(IBase.*)
around call(IBase+.*)
IBase.base1 from Intro
around call(IBase.*)
around call(IBase+.*)
around execution(Base+.*)
Base.base2
around call(IBase.*)
around call(IBase+.*)
IBase.base3 from Intro
**************************************
around call(IBase.*)
around call(IBase+.*)
IDerived.base1 from Intro
around call(IBase.*)
around call(IBase+.*)
IDerived.base2 from Intro
around call(IBase.*)
around call(IBase+.*)
Derived.base3
**************************************
around call(IBase.*)
around call(IBase+.*)
IBase.base1 from Intro
around call(IBase.*)
around call(IBase+.*)
around execution(Base+.*)
Base.base2
around call(IBase.*)
around call(IBase+.*)
around execution(Base+.*)
SubBase.base3
